pygfx/renderers/wgpu/wgsl/image_common.wgsl

```diff
@@ -2,10 +2,57 @@
 
 {$ include 'pygfx.image_sample.wgsl' $}
 
+fn yuv_limited_to_rgb(y: f32, u: f32, v: f32) -> vec4<f32> {
+    // This formula is correct for the "limited range" YUV
+    let c: f32 = y - 0.0625;        // Offset Y by 16/255
+    let d: f32 = u - 0.5;           // Offset U by 128/255
+    let e: f32 = v - 0.5;           // Offset V by 128/255
+
+    let r: f32 = 1.1643 * c + 1.5958 * e;
+    let g: f32 = 1.1643 * c - 0.3917 * d - 0.8129 * e;
+    let b: f32 = 1.1643 * c + 2.0170 * d;
+
+    return vec4<f32>(r, g, b, 1.);
+}
+
+fn yuv_full_to_rgb(y: f32, u: f32, v: f32) -> vec4<f32> {
+    // this formula is correct for the "full range" YUV
+    let d: f32 = u - 0.5;           // Offset U by 128/255
+    let e: f32 = v - 0.5;           // Offset V by 128/255
+
+    let r = y + 1.402 * e;
+    let g = y - 0.344136 * d - 0.714136 * e;
+    let b = y + 1.772 * d;
+
+    return vec4<f32>(r, g, b, 1.);
+}
 
 fn sample_im(texcoord: vec2<f32>, sizef: vec2<f32>) -> vec4<f32> {
     $$ if img_format == 'f32'
-        return textureSample(t_img, s_img, texcoord.xy);
+        $$ if colorspace == 'yuv420p'
+            let y = textureSample(t_img, s_img, texcoord.xy, 0).x;
+            let u = textureSample(t_img, s_img, texcoord.xy / 2.0, 1).x;
+            let v = textureSample(t_img, s_img, texcoord.xy / 2.0 + vec2<f32>(0.5, 0.0), 1).x;
```

Comment:
I'm not 100% convinced you avoid a copy when the source is contiguous and the destination is not contiguous.

Your destination buffer here implies that you are using:

```
   ________
  |         |
  |  y      |
  |_________|
  | u  |    |
  |---------|
  |_v__|____|
```
Which is not super optimal.

Reply:
The v is actually to the right of the v (I changed that relatively recently, but it does not really matter). In any case, what you say is true: in the destination texture, the u and v are not contiguous. I don't think this is a problem, it does not result in an extra copy or anything like that, it's copying the data into the texture anyway, and the API for doing so does not suggest that strides hurt performance.

It would be possible to copy u and v to a contiguous part of the texture, but then the interpolation would be all wrong.

Reply:
Btw. You are interpolating between u and v at their boundary as is. So it isn’t entirely correct as is right now.

Reply:
Good point, we may want to tweak the texcoords to clip the last half-pixel.

Reply:
i'm not sure you'll be able to do it for all the different interpolation modes. How would you get around bicubic???? Is that a hardware interpolation mode?

In either case, with the 3 grid mode, I'm not sure it is strictly necessary to do it for the "3D grid" mode.

Reply:
> i feel like you have a good idea on how to do this. Can i leave it to you?

Yeah, I can have a look.

> How would you get around bicubic???? Is that a hardware interpolation mode?

Anything beyond linear interpolation (including cubic) must be implemented in the shader.